presto-iceberg/src/main/java/io/prestosql/iceberg/IcebergMetadata.java

```diff
@@ -0,0 +1,612 @@
+    private final HdfsEnvironment hdfsEnvironment;
+    private final TypeManager typeManager;
+    private final SemiTransactionalHiveMetastore metastore;
+    private final JsonCodec<CommitTaskData> jsonCodec;
+    private Transaction transaction;
+    private IcebergUtil icebergUtil;
+    private final LocationService locationService;
+
+    public IcebergMetadata(
+            SemiTransactionalHiveMetastore metastore,
+            HdfsEnvironment hdfsEnvironment,
+            TypeManager typeManager,
+            JsonCodec<CommitTaskData> jsonCodec,
+            IcebergUtil icebergUtil,
+            LocationService locationService)
+    {
+        this.hdfsEnvironment = hdfsEnvironment;
+        this.typeManager = typeManager;
+        this.metastore = metastore;
+        this.jsonCodec = jsonCodec;
+        this.icebergUtil = icebergUtil;
+        this.locationService = locationService;
+    }
+
+    @Override
+    public List<String> listSchemaNames(ConnectorSession session)
+    {
+        return metastore.getAllDatabases();
+    }
+
+    @Override
+    public IcebergTableHandle getTableHandle(ConnectorSession session, SchemaTableName tableName)
+    {
+        Optional<Table> table = metastore.getTable(tableName.getSchemaName(), tableName.getTableName());
+        if (table.isPresent()) {
+            if (icebergUtil.isIcebergTable(table.get())) {
+                return new IcebergTableHandle(tableName.getSchemaName(), tableName.getTableName());
+            }
+            else {
+                throw new RuntimeException(String.format("%s is not an iceberg table please query using hive catalog", tableName));
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public Optional<SystemTable> getSystemTable(ConnectorSession session, SchemaTableName tableName)
+    {
+        if (!isPartitionsSystemTable(tableName)) {
+            return Optional.empty();
+        }
+
+        SchemaTableName sourceTableName = getSourceTableNameForPartitionsTable(tableName);
+        String schemaName = sourceTableName.getSchemaName();
+        IcebergTableHandle sourceTableHandle = getTableHandle(session, sourceTableName);
+        List<HiveColumnHandle> partitionColumns = getColumnHandles(session, sourceTableHandle)
+                .entrySet().stream().filter(e -> ((HiveColumnHandle) e.getValue()).isPartitionKey())
+                .map(e -> (HiveColumnHandle) e.getValue())
+                .collect(Collectors.toList());
+
+        if (partitionColumns == null || partitionColumns.isEmpty()) {
+            return Optional.empty();
+        }
+
+        List<ColumnMetadata> columnMetadatas = partitionColumns.stream()
+                .map(columnHandle -> getColumnMetadata(session, sourceTableHandle, columnHandle))
+                .collect(Collectors.toList());
+        List<String> partitionColumnNames = Lists.newArrayList("record_count", "file_count", "total_size");
```

Comment:
Use `new ArrayList<>()`. These factories are legacy from before Java had the diamond operator.